File: DSA/Aanidhay-Aggarwal/Day-18/Day18_q1.java
import java.util.*;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}

public class Main {
    public static List<Integer> topView(TreeNode root) {
        if (root == null) {
            return new ArrayList<>();
        }

        Map<Integer, Integer> map = new HashMap<>();
        Queue<TreeNode> queue = new LinkedList<>();
        Queue<Integer> horizontalDistanceQueue = new LinkedList<>();

        queue.add(root);
        horizontalDistanceQueue.add(0);

        while (!queue.isEmpty()) {
            TreeNode node = queue.poll();
            int horizontalDistance = horizontalDistanceQueue.poll();

            if (!map.containsKey(horizontalDistance)) {
                map.put(horizontalDistance, node.val);
            }

            if (node.left!= null) {
                queue.add(node.left);
                horizontalDistanceQueue.add(horizontalDistance - 1);
            }

            if (node.right!= null) {
                queue.add(node.right);
                horizontalDistanceQueue.add(horizontalDistance + 1);
            }
        }

        List<Integer> result = new ArrayList<>();
        for (int i = Collections.min(map.keySet()); i <= Collections.max(map.keySet()); i++) {
            result.add(map.get(i));
        }

        return result;
    }

    public static void main(String[] args) {
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(3);
        root.left.left = new TreeNode(4);
        root.left.right = new TreeNode(5);
        root.right.left = new TreeNode(6);
        root.right.right = new TreeNode(7);

        List<Integer> result = topView(root);
        System.out.println(result);
    }
}

File: DSA/Hima-Swarupa-Padala/day18q1.py
class TreeNode:  
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None

def top_view(root):
    if not root:
        return []

    hd_dict = {}
    node_dict = {}

    def traverse(node, hd, level):       
        if not node:
            return            
        if hd not in hd_dict or level < hd_dict[hd]:
            hd_dict[hd] = level
            node_dict[hd] = node.value
        traverse(node.left, hd - 1, level + 1)
        traverse(node.right, hd + 1, level + 1)

    traverse(root, 0, 0)

    sorted_nodes = [node_dict[hd] for hd in sorted(hd_dict.keys())]

    return sorted_nodes

if __name__ == "__main__":

    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    root.left.left = TreeNode(4)
    root.left.right = TreeNode(5)
    root.right.left = TreeNode(6)
    root.right.right = TreeNode(7)

    top_view_nodes = top_view(root)    
    print(top_view_nodes)
